### DISCUSSION TITLE
Are the modules something important to start a nuxt3?

### DISCUSSION THREAD

**Discussion by saifobeidat** [nuxt/nuxt#18064] [Questions]
Hello,

I am new to Nuxt, my company is welling to move an existing project to Nuxt3.. we see that nuxt3 is still beta and it's not clear when will it be released.. in our case the the modules support https://modules.nuxtjs.org/ is the only major thing that's missing.

**Our project will have these modules:**
- i18
- apollo  
- Google Analytics
- Tailwind v3

and more..

I have two questions:
- Is it necessary to have the above modules supported in Nuxt3 in order to use them in a nuxt3 project?
- What are the benefits of using supported modules? I mean why not just install the main vue or Js module to the project, like using this https://kazupon.github.io/vue-i18n/ instead of nuxt/i18  what the difference will be?

thanks

**Comment by wobsoriano** [comment]
> Is it necessary to have the above modules supported in Nuxt3 in order to use them in a nuxt3 project?

Yes, most of them.

> What are the benefits of using supported modules?

Nuxt3 uses vite. Most of modules will fail because of esm support

**Reply by saifobeidat** [reply]
Thanks @wobsoriano 
Okay, but Nuxt at the end is a vue project, like for tailwind it's available for Vue3
https://tailwindcss.com/docs/guides/vite
isn't possible to install it to a nuxt project without having it supported in Nuxt? and if yes, how to install external package to Nuxt..

thanks :)

**Reply by wobsoriano** [reply]
You can always use whatever is available if it works

**Reply by arb85** [reply]
> Thanks @wobsoriano 
> 
> Okay, but Nuxt at the end is a vue project, like for tailwind it's available for Vue3
> 
> https://tailwindcss.com/docs/guides/vite
> 
> isn't possible to install it to a nuxt project without having it supported in Nuxt? and if yes, how to install external package to Nuxt..
> 
> 
> 
> thanks :)  

If you don't really need ALL the features of Tailwind V3 you can give it a try to WindiCSS, which is pretty similar (with a lot of other features) and you can use the same Tailwind syntax

NUXT 3 Module here:
https://github.com/windicss/nuxt-windicss

**Comment by sewalsh** [comment]
>    What are the benefits of using supported modules? I mean why not just install the main vue or Js module to the project, like using this https://kazupon.github.io/vue-i18n/ instead of nuxt/i18  what the difference will be?

I consider modules as an easy way of adding a new library without having to configure it as a plugin. For sure there's always the possibility of integrating a library as plugin.

**Comment by viii47** [comment]
> * What are the benefits of using supported modules? I mean why not just install the main vue or Js module to the project, like using this https://kazupon.github.io/vue-i18n/ instead of nuxt/i18  what the difference will be?

Do you get benefits from using a supported module for nuxt? Definitely. Do you have to? Yes** and no.

A lot of modules that are not specifically made for nuxt might work just out-of-the-box right off, others will need some modifications and some others need new hooks, handlers, overhead code to even be used. It depends on what the goal is with the extra functionality of a plugin or module. 

**This is more of an elemental question for yourself to answer, also what kind of functionality you expect out of a specific module. You could just inject the code in the mounted section and call it a day, but then you would lose all the SSR benefits.

----
Personally I do both. I use robust and *tested by time* modules that seem safe to use. Nuxt3 is still not *production-ready* so expecting having everything carte blanche is unlikely. I started a new project that is meant for a production-use in the near future. So while looking around I wasn't satisfied with some things. 

A self-made module - my own i18n plugin that works *for my purposes* wonderfully; a simple plugin with utility code to read a route param to know what i18n-json to use `e.g. http://yolo.com/en/welcome/... -> ../locales/en-us.json`, setting the meta headers correctly for both, SSR and client, injecting a declared interface to use it in every component, plugin, ..etc. I still need to make some adjustments since I want to preload some json files and some failsafes. 

For some things I'm just too lazy to reinvent the wheel. Such a case is a gallery/swiping plugin. I was looking around and decided to have a run at `swiperJS` which does have a nuxt specific module written for it but I didn't fancy some things around it so I tuned the crazy up a notch and tried to implement it myself. Checked out the documentation and it's actually really easy to get it rolling. 
- npm install the required assets
- make two new components `Slider.vue, SliderItem.vue`
- import the core modules
- make some fancy props and code
- call it a day

It still has some bugs that I need to iron out but so far it works as intended. An excerpt of the component:
```js
<template>
    <!-- Slider main container -->
    <div
        class="swiper h-auto min-h-100px max-h-400px swiper__notactive"
        :style="`height:${height}`"
        :class="`swiper-index-${$props.name}`"
    >
        <div class="swiper-wrapper">
            <!-- Slides -->
            <slot></slot>
        </div>

        <!-- If we need pagination -->
        <div class="swiper-pagination"></div>

        <!-- If we need navigation buttons -->
        <div class="swiper-button-prev" v-if="$props.navigation"></div>
        <div class="swiper-button-next" v-if="$props.navigation"></div>
    </div>
</template>

<script lang="ts">
// core version + navigation, pagination modules:
import { Swiper, Pagination, Autoplay, EffectFade } from 'swiper';
import { logdebug, logerror } from '@/lib/util/log';
export default defineComponent({
    props: {
        direction: {
            type: String,
            required: false,
            default: 'horizontal'
        },
        pagination: {
            type: Boolean,
            required: false,
            default: true
        },
        autoplay: {
            type: Boolean,
            required: false,
            default: false
        },
        height: {
            type: String,
            required: false,
            default: "100px"
        },
        fade: {
            type: Boolean,
            required: false,
            default: false
        },
        delay: {
            type: Number,
            required: false,
            default: 4500
        },
        slidesPerView: {
            type: Number,
            required: false,
            default: 1
        },
        name: {
            type: String,
            required: false,
            default: 'default'
        },
        navigation: {
            type: Boolean,
            required: false,
            default: false
        }
    },
    mounted() {
        setTimeout(() => {
            Swiper.use([Pagination, Autoplay, EffectFade]);
            const swiper = new Swiper(`.swiper-index-${this.$props.name}`, {
                // Optional parameters
                direction: this.direction,
                loop: true,
                // If we need pagination
                pagination: {
                    el: '.swiper-pagination'
                },
                slidesPerView: this.slidesPerView,
                spaceBetween: 30,
                effect: this.fade ? 'fade' : 'slide',
                fadeEffect: {
                    crossFade: this.fade ? true : false
                },
            });
            if (this.autoplay) {
                swiper.params.autoplay = { delay: this.delay, disableOnInteraction: false };
                swiper.autoplay.start();
            }
            try {
                const verticalPaginator = document.getElementsByClassName('swiper-pagination-horizontal swiper-pagination-vertical')[0];
                const horizontalPaginator = document.getElementsByClassName('swiper-pagination-vertical swiper-pagination-horizontal')[0];
                if (this.direction === 'vertical' && verticalPaginator !== undefined) {
                    verticalPaginator.className = 'swiper-pagination swiper-pagination-bullets swiper-pagination-vertical';
                }
                if (this.direction === 'horizontal' && horizontalPaginator !== undefined) {
                    horizontalPaginator.className = 'swiper-pagination swiper-pagination-bullets swiper-pagination-horizontal';
                }
            } catch (error) {
                logerror('Slider -> mounted()', error);
            }
            logdebug('Slider -> mounted()', `swiper initialized as swiper-${this.$props.name}`);
        }, 150);
    }
})

</script>
```

Conclusion: If a module exists for nuxt - give it definitely a try. If it has some bugs, try to work around it - if not, try to use the original/core version for vue/default/core/... and use that instead. Or go even beyond and write your completely self made code. It's really up to you.
